debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/events/BufferingChangeStreamCursor.java

```diff
@@ -0,0 +1,303 @@
+        private final AtomicBoolean running;
+        private final AtomicReference<MongoChangeStreamCursor<ChangeStreamDocument<TResult>>> cursorRef;
+        private final MongoDbStreamingChangeEventSourceMetrics metrics;
+        private final Clock clock;
+        private int noMessageIterations = 0;
+
+        public EventFetcher(ChangeStreamIterable<TResult> stream,
+                            int capacity,
+                            DelayStrategy pause,
+                            MongoDbStreamingChangeEventSourceMetrics metrics,
+                            Clock clock) {
+            this.stream = stream;
+            this.capacity = new Semaphore(capacity);
+            this.pause = pause;
+            this.metrics = metrics;
+            this.clock = clock;
+            this.running = new AtomicBoolean(false);
+            this.cursorRef = new AtomicReference<>(null);
+            this.queue = new ConcurrentLinkedQueue<>();
+        }
+
+        /**
+         * Indicates whether event fetching is running and the internal cursor is open
+         *
+         * @return true if running, false otherwise
+         */
+        public boolean isRunning() {
+            return running.get();
+        }
+
+        @Override
+        public void close() {
+            running.set(false);
+        }
+
+        public ResumableChangeStreamEvent<TResult> poll() {
+            var event = queue.poll();
+            if (event != null) {
+                capacity.release();
+            }
+            return event;
+        }
+
+        public boolean isEmpty() {
+            return queue.isEmpty();
+        }
+
+        /**
+         * Depending on queue implementation this method may not be reliable
+         * By default see {@link ConcurrentLinkedQueue#size()}
+         *
+         * @return approximate number of elements in queue
+         */
+        public int size() {
+            return queue.size();
+        }
+
+        @Override
+        public void run() {
+            try (MongoChangeStreamCursor<ChangeStreamDocument<TResult>> cursor = stream.cursor()) {
+                cursorRef.compareAndSet(null, cursor);
+                running.set(true);
+                noMessageIterations = 0;
+                fetchEvents(cursor);
+            }
+            catch (InterruptedException e) {
+                throw new DebeziumException("Fetcher thread interrupted", e);
+            }
+            finally {
+                cursorRef.set(null);
+                close();
+            }
+        }
+
+        private void fetchEvents(MongoChangeStreamCursor<ChangeStreamDocument<TResult>> cursor) throws InterruptedException {
+            while (isRunning()) {
+                var beforeEventPollTime = clock.currentTimeAsInstant();
+                var event = new ResumableChangeStreamEvent<>(cursor.next(), cursor);
+                event.document.ifPresent(doc -> LOGGER.trace("Polled Change Stream event: {}", event));
```

Comment:
this is probably not optimal as we should ensure that ResumableEvents are only enqueued when the presence of resume token is guaranteed (it might not be in some "Mongo compatible" databases and in case of bugs like SERVER-63772) 

in such case the time diff might be captured as higher. 

@jpechane  also please remind me which of the metrics we wanted to kept? poll time for event / no event, count of no events? Count of events as added recently is duplicit IMHO and should be removed.